### ISSUE TITLE
Windows doesn't build out-of-the-box

### ISSUE THREAD

**Issue by Timmmm** [gtk-rs/gtk#494]
Hi,

I downloaded and installed MSYS2 with the default path (`C:\msys64`) and then ran `pacman -S mingw-w64-x86_64-gtk3` as directed on the [GTK+ download page](https://www.gtk.org/download/windows.php) (why they don't have a proper Windows installer... smh). Anyway that seemed to work. I then select what I believe should be the right rust version:

    rustup default stable-i686-pc-windows-gnu

And then `cargo build` an example. However sadly it can't link to gtk:

error: linking with `gcc` failed: exit code: 1
  |
```
  = note: "gcc" "-Wl,--enable-long-section-names" "-fno-use-linker-plugin" "-Wl,--nxcompat" "-nostdlib" "-m64" "C:\\Users\\Tim\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\crt2.o" "C:\\Users\\Tim\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsbegin.o" "-L" "C:\\Users\\Tim\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\ViewSB-4e41fd1abce93757.0.o" "-o" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\ViewSB-4e41fd1abce93757.exe" "-Wl,--gc-sections" "-nodefaultlibs" "-L" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps" "-L" "C:\\Users\\Tim\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "-Wl,-Bstatic" "-Wl,-Bdynamic" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libgtk-c525a777fdab89f4.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libgtk_sys-bae90cffcf3ef07a.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libgio-d0b6c354a4a6032f.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libpango-62d733263b531c0d.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libatk_sys-4c92419d575bce35.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libgdk-bc3c4e717db5a6b8.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libcairo-6e0780fdc7c8779a.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libgdk_pixbuf-57e2a05c8d3823fa.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libglib-fea96c909af070b7.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libbitflags-8510a47beebe00a6.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libgdk_sys-1c5dbb0b6e029628.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libcairo_sys-9adc0cdd404adc66.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libwinapi-a5898d7aceb63fac.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libpango_sys-0dc6768add43888b.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libgdk_pixbuf_sys-55f316a12338ec76.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libgio_sys-6dbba2ca0ce319da.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libgobject_sys-4ba000a4d9f2cf78.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libglib_sys-895440df8bd9794c.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libbitflags-6c0c42f1f858d67c.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\liblibc-5dc7b85e748840b4.rlib" "C:\\Users\\Tim\\Dropbox\\Projects\\ViewSB\\target\\debug\\deps\\libc_vec-3ad5b18b850a7482.rlib" "C:\\Users\\Tim\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libstd-2534c0bd05ee265e.rlib" "C:\\Users\\Tim\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libpanic_unwind-e3cf511e7374526b.rlib" "C:\\Users\\Tim\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libunwind-52e8ab7efa40b1b3.rlib" "C:\\Users\\Tim\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\liblibc-96a4c6f6006d0a44.rlib" "C:\\Users\\Tim\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\librand-32b552aabc8b4e3c.rlib" "C:\\Users\\Tim\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libcollections-4965c4d8d37432ab.rlib" "C:\\Users\\Tim\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\liballoc-8bdc91ebec92bb48.rlib" "C:\\Users\\Tim\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\liballoc_system-1769b774bd375392.rlib" "C:\\Users\\Tim\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libstd_unicode-feba27dd5784a3dc.rlib" "C:\\Users\\Tim\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libcore-d1dd5a39fb635c62.rlib" "C:\\Users\\Tim\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libcompiler_builtins-35b663261c982c37.rlib" "-l" "gtk-3" "-l" "gdk-3" "-l" "atk-1.0" "-l" "gdk-3" "-l" "cairo" "-l" "pango-1.0" "-l" "gdk_pixbuf-2.0" "-l" "gio-2.0" "-l" "gobject-2.0" "-l" "glib-2.0" "-l" "gobject-2.0" "-l" "advapi32" "-l" "ws2_32" "-l" "userenv" "-l" "shell32" "-l" "gcc_eh" "-lmingwex" "-lmingw32" "-lgcc" "-lmsvcrt" "-luser32" "-lkernel32" "C:\\Users\\Tim\\.rustup\\toolchains\\stable-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsend.o"
  = note: ld: cannot find -lgtk-3
          ld: cannot find -lgdk-3
          ld: cannot find -latk-1.0
          ld: cannot find -lgdk-3
          ld: cannot find -lcairo
          ld: cannot find -lpango-1.0
          ld: cannot find -lgdk_pixbuf-2.0
          ld: cannot find -lgio-2.0
          ld: cannot find -lgobject-2.0
          ld: cannot find -lglib-2.0
          ld: cannot find -lgobject-2.0

error: aborting due to previous error

error: Could not compile `ViewSB`.
```

I *think* I haven't done anything wrong. Maybe it just isn't given the right path, but I feel like it should at least build out of the box if you install everything with the default paths.

**Comment by GuillaumeGomez** [comment]
You need to have the libs in your path. If `rustc` cannot find them, then you failed somewhere.

**Comment by Timmmm** [comment]
Yeah, but I mean it would be nice if it worked without me having to set the `PATH` and `GTK_LIB_DIR`. The default could be to use the default MSYS install directory.

**Comment by Geobert** [comment]
I have the same issue, event after setting GTK_LIB_DIR and PATH accordingly to http://gtk-rs.org/docs-src/requirements

What am I missing here? :-/

**Comment by Geobert** [comment]
I was in feature 3_10 changed to 3_16 (just because I've seen it on http://gtk-rs.org/docs-src/) make the compilation works Oo

**Comment by GuillaumeGomez** [comment]
Features are important. :)

**Comment by EPashkin** [comment]
But I don't see how wrong feature can produce this error ☹️

**Comment by Geobert** [comment]
Maybe update the homepage to reflect the min feature needed to have it compiles :)

**Comment by GuillaumeGomez** [comment]
@EPashkin: Me neither. :-/

**Comment by Susurrus** [comment]
@Geobert I have the same issue but upping the feature didn't resolve it.

As I understand it, I'm on 32-bit Windows 10. I installed msys2 (actually to `C:\msys32` which is the default, not `C:\msys2`). I opened the MSYS2 MinGW 32-bit window and installed both the `mingw-w64-i686-gtk3` and `mingw-w64-i686-toolchain` (selecting the default of "all"). Linking fails with the same errors as posted by @Timmmm. Do note that I didn't explicitly set any environment variables, as the Requirements page doesn't say that's necessary if you install the toolchain in mingw. Is there a step I'm missing?

**Comment by EPashkin** [comment]
@Susurrus How you installed rust?
What says `pkg-config --modversion gtk+-3.0` in mingw32bit console?

**Comment by Susurrus** [comment]
I installed Rust via rustup.

Running your suggested command I see that `pkg-config` couldn't find the libraries even though I already installed them (running `pacman -S mingw-w64-i686-gtk3` says `-- reinstalling` suggesting that this is the case).

```
$ pkg-config --modversion gtk+3.0
Package gtk+3.0 was not found in the pkg-config search path.
Perhaps you should add the directory containing `gtk+3.0.pc'
to the PKG_CONFIG_PATH environment variable
No package 'gtk+3.0' found
```

**Comment by EPashkin** [comment]
Installed rust in same mingw32bit console?

I forgot `-` in command line: pkg-config --modversion gtk+-3.0

**Comment by EPashkin** [comment]
pkg-config --libs gtk+-3.0
Shows right path?

**Comment by EPashkin** [comment]
If pkg-config don't see it, maybe problem that you installed packages in MINGW32 console not in MSYS2.

**Comment by EPashkin** [comment]
Strange I can't install rust on windows7 VM.
I always get `error: error parsing settings` even in default installation on https://sh.rustup.rs or rustup-init (even tried 64x version)

**Comment by EPashkin** [comment]
Still can't work with rust in msys32 command line, it can't run cargo or rustc even if it in current folder.

**Comment by EPashkin** [comment]
Steps that seems work on empty Windows7 VM
1. Install rust with rustup‑init.exe (as script in console install to "C:\users\USER" anyway)
with i686-pc-windows-gnu/stable/yes
2. Install MSYS2 from http://www.msys2.org (used  i686 version but in add 64console anyway)
I do use link different from http://gtk-rs.org/docs-src/requirements.html as sourceforge.net asked login
3. Added `PATH="/c/Users/${USER}/.cargo/bin:${PATH}"` to `C:\MSYS32\home\USER\.bash_profile`
4. In MSYS2 MSYS console
```
pacman -S git
pacman -S mingw-w64-i686-gtk3
pacman -S mingw-w64-i686-toolchain
default parameters
```
5. In MSYS2 MinGW 32-bit console
```
git clone https://github.com/gtk-rs/examples/
cd examples
cargo build
cargo run --bin basic
```

**Comment by EPashkin** [comment]
@Susurrus you found problem on your end?

**Comment by Susurrus** [comment]
Not quite. I followed your instructions. Running `pkg-config --modversion gtk+-3.0` in the MSYS2 MinGW 32-bit console gives me `3.22.1`. Now running the example code works following your instructions. Running my own project, however, fails with the same `ld: cannot find -lgtk-3 ...` until I did a `cargo clean`, so now it's all good.

Don't really know what was the problem before, but looks like it's fixed now. Thanks, @EPashkin!

**Comment by mati865** [comment]
@Susurrus I suspect you had `msys/gcc` installed but no `mingw32/mingw-w64-gcc` and POSIX emulated libraries got mixed with Windows native libraries during build.
`cargo clean` removed miscompiled code.

**Comment by thatcodeguy** [comment]
You're trying to use the `stable-i686-pc-windows-gnu` target and the `mingw-w64-x86_64-gtk3` package.

Have you tried installing `mingw-w64-i686-gtk3` (`pacman -S mingw-w64-i686-gtk3`)?

**Comment by ctaggart** [comment]
I haven't been able to get the basic example running on Windows using either 32-bit or 64-bit. I got it building in both places, but it fails at runtime. I've been using `ldd` to troubleshoot what appears to be dependency problems. I tried following @EPashkin's instructions. Could someone who has it working give me the output from `ldd target/debug/basic.exe`? Here is what I've got on my system for 32-bit:

```
camer@delltop MINGW32 /c/Users/camer/rs/gtk-rs-examples
$ cargo run --bin basic
    Finished dev [unoptimized + debuginfo] target(s) in 0.1 secs
     Running `target\debug\basic.exe`
error: process didn't exit successfully: `target\debug\basic.exe` (exit code: 3221225785)


camer@delltop MINGW32 /c/Users/camer/rs/gtk-rs-examples
$ rustup show
Default host: x86_64-pc-windows-msvc

installed toolchains
--------------------

stable-i686-pc-windows-gnu (default)
stable-x86_64-pc-windows-gnu
stable-x86_64-pc-windows-msvc
beta-x86_64-pc-windows-gnu
nightly-x86_64-pc-windows-gnu
nightly-x86_64-pc-windows-msvc

active toolchain
----------------

stable-i686-pc-windows-gnu (default)
rustc 1.19.0 (0ade33941 2017-07-17)


camer@delltop MINGW32 /c/Users/camer/rs/gtk-rs-examples
$ pkg-config --modversion gtk+-3.0
3.22.16


camer@delltop MINGW32 /c/Users/camer/rs/gtk-rs-examples
$ ldd target/debug/basic.exe
        ntdll.dll => /c/WINDOWS/SYSTEM32/ntdll.dll (0x7ff84c370000)
        ??? => ??? (0x77340000)
        wow64.dll => /c/WINDOWS/System32/wow64.dll (0x561a0000)
        wow64win.dll => /c/WINDOWS/System32/wow64win.dll (0x56120000)
```

**Comment by GuillaumeGomez** [comment]
You should maybe give a try to dependency-walker. I find this tool very nice when working on windows.

**Comment by ctaggart** [comment]
I tried  dependency-walker and procmon and wasn't able to diagnose the problem. Having the output from a working system would be really helpful for:

```
$ pkg-config --modversion gtk+-3.0
```
and
```
$ ldd target/debug/basic.exe
```

**Comment by mati865** [comment]
pkg-config will be the same on both systems.
You can check nm and objdump but first try with higher feature level in project config.

**Comment by EPashkin** [comment]
```
$ pkg-config --modversion gtk+-3.0
3.22.15

$ pkg-config --libs gtk+-3.0
-LC:/msys32/mingw32/lib -lgtk-3 -lgdk-3 -lgdi32 -limm32 -lshell32 -lole32 -Wl,-luuid -lwinmm -ldwmapi -lsetupapi -lcfgmgr32 -lz -lpangowin32-1.0 -lpangocairo-1.0 -lpango-1.0 -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lintl

$ ldd target/debug/basic.exe
        ntdll.dll => /c/Windows/SysWOW64/ntdll.dll (0x77bd0000)
        kernel32.dll => /c/Windows/syswow64/kernel32.dll (0x75e20000)
        KERNELBASE.dll => /c/Windows/syswow64/KERNELBASE.dll (0x77110000)
        libcairo-2.dll => /mingw32/bin/libcairo-2.dll (0x68fc0000)
        GDI32.dll => /c/Windows/syswow64/GDI32.dll (0x76fb0000)
        USER32.dll => /c/Windows/syswow64/USER32.dll (0x77160000)
        ADVAPI32.dll => /c/Windows/syswow64/ADVAPI32.dll (0x75fe0000)
        msvcrt.dll => /c/Windows/syswow64/msvcrt.dll (0x75f30000)
        sechost.dll => /c/Windows/SysWOW64/sechost.dll (0x75990000)
        RPCRT4.dll => /c/Windows/syswow64/RPCRT4.dll (0x77440000)
        SspiCli.dll => /c/Windows/syswow64/SspiCli.dll (0x75730000)
        CRYPTBASE.dll => /c/Windows/syswow64/CRYPTBASE.dll (0x75720000)
        LPK.dll => /c/Windows/syswow64/LPK.dll (0x75e10000)
        USP10.dll => /c/Windows/syswow64/USP10.dll (0x76f10000)
        MSIMG32.DLL => /c/Windows/system32/MSIMG32.DLL (0x73e00000)
        libgcc_s_dw2-1.dll => /mingw32/bin/libgcc_s_dw2-1.dll (0x6eb40000)
        libwinpthread-1.dll => /mingw32/bin/libwinpthread-1.dll (0x64b40000)
        libfontconfig-1.dll => /mingw32/bin/libfontconfig-1.dll (0x65180000)
        libexpat-1.dll => /mingw32/bin/libexpat-1.dll (0x69140000)
        libfreetype-6.dll => /mingw32/bin/libfreetype-6.dll (0x62d40000)
        libbz2-1.dll => /mingw32/bin/libbz2-1.dll (0x628c0000)
        libharfbuzz-0.dll => /mingw32/bin/libharfbuzz-0.dll (0x61800000)
        libglib-2.0-0.dll => /mingw32/bin/libglib-2.0-0.dll (0x687c0000)
        ole32.dll => /c/Windows/syswow64/ole32.dll (0x77530000)
        SHELL32.dll => /c/Windows/syswow64/SHELL32.dll (0x76100000)
        SHLWAPI.dll => /c/Windows/syswow64/SHLWAPI.dll (0x76d60000)
        WINMM.DLL => /c/Windows/system32/WINMM.DLL (0x743a0000)
        WS2_32.dll => /c/Windows/syswow64/WS2_32.dll (0x77040000)
        NSI.dll => /c/Windows/syswow64/NSI.dll (0x77690000)
        libintl-8.dll => /mingw32/bin/libintl-8.dll (0x61ec0000)
        libiconv-2.dll => /mingw32/bin/libiconv-2.dll (0x66200000)
        libpcre-1.dll => /mingw32/bin/libpcre-1.dll (0x69340000)
        libgraphite2.dll => /mingw32/bin/libgraphite2.dll (0x70740000)
        libstdc++-6.dll => /mingw32/bin/libstdc++-6.dll (0x6fe40000)
        libpng16-16.dll => /mingw32/bin/libpng16-16.dll (0x68d40000)
        zlib1.dll => /mingw32/bin/zlib1.dll (0x63080000)
        libpixman-1-0.dll => /mingw32/bin/libpixman-1-0.dll (0x65500000)
        libgdk-3-0.dll => /mingw32/bin/libgdk-3-0.dll (0x710c0000)
        dwmapi.dll => /c/Windows/system32/dwmapi.dll (0x73f50000)
        IMM32.DLL => /c/Windows/syswow64/IMM32.DLL (0x759b0000)
        MSCTF.dll => /c/Windows/syswow64/MSCTF.dll (0x77700000)
        SETUPAPI.dll => /c/Windows/syswow64/SETUPAPI.dll (0x75790000)
        CFGMGR32.dll => /c/Windows/syswow64/CFGMGR32.dll (0x76dc0000)
        OLEAUT32.dll => /c/Windows/syswow64/OLEAUT32.dll (0x77080000)
        DEVOBJ.dll => /c/Windows/syswow64/DEVOBJ.dll (0x75df0000)
        libcairo-gobject-2.dll => /mingw32/bin/libcairo-gobject-2.dll (0x6d140000)
        libgobject-2.0-0.dll => /mingw32/bin/libgobject-2.0-0.dll (0x63c40000)
        libffi-6.dll => /mingw32/bin/libffi-6.dll (0x6b940000)
        libepoxy-0.dll => /mingw32/bin/libepoxy-0.dll (0x6e100000)
        libgdk_pixbuf-2.0-0.dll => /mingw32/bin/libgdk_pixbuf-2.0-0.dll (0x3a0000)
        libgio-2.0-0.dll => /mingw32/bin/libgio-2.0-0.dll (0x6d780000)
        libgmodule-2.0-0.dll => /mingw32/bin/libgmodule-2.0-0.dll (0x6df00000)
        DNSAPI.dll => /c/Windows/system32/DNSAPI.dll (0x73e20000)
        IPHLPAPI.DLL => /c/Windows/system32/IPHLPAPI.DLL (0x73de0000)
        WINNSI.DLL => /c/Windows/system32/WINNSI.DLL (0x73dd0000)
        gdiplus.dll => /c/Windows/WinSxS/x86_microsoft.windows.gdiplus_6595b64144ccf1df_1.1.7601.17514_none_72d18a4386696c80/gdiplus.dll (0x73c40000)
        libpango-1.0-0.dll => /mingw32/bin/libpango-1.0-0.dll (0x65780000)
        libpangocairo-1.0-0.dll => /mingw32/bin/libpangocairo-1.0-0.dll (0x6d6c0000)
        libpangoft2-1.0-0.dll => /mingw32/bin/libpangoft2-1.0-0.dll (0x3e0000)
        libpangowin32-1.0-0.dll => /mingw32/bin/libpangowin32-1.0-0.dll (0x6b480000)
        libgtk-3-0.dll => /mingw32/bin/libgtk-3-0.dll (0x66600000)
        COMCTL32.DLL => /c/Windows/WinSxS/x86_microsoft.windows.common-controls_6595b64144ccf1df_6.0.7601.17514_none_41e6975e2bd6f2b2/COMCTL32.DLL (0x74090000)
        COMDLG32.DLL => /c/Windows/syswow64/COMDLG32.DLL (0x76080000)
        WINSPOOL.DRV => /c/Windows/system32/WINSPOOL.DRV (0x74030000)
        libatk-1.0-0.dll => /mingw32/bin/libatk-1.0-0.dll (0x62b40000)
        USERENV.dll => /c/Windows/system32/USERENV.dll (0x73c20000)
        profapi.dll => /c/Windows/system32/profapi.dll (0x73c10000)
```

**Comment by EPashkin** [comment]
```
$ whereis libgtk-3-0
libgtk-3-0: /mingw32/bin/libgtk-3-0.dll
```

**Comment by ctaggart** [comment]
Thanks @EPashkin! I've got it working now in 64-bit mode. :grin:

``` powershell
rustup default stable-gnu
$env:path = "C:\msys64\mingw64\bin;$env:path"
cargo run --bin basic
```

Launching the process from PowerShell really helped diagnose the problems. Anyone having trouble, I'd highly recommend trying this:
``` powershell
Start-Process -PassThru target\debug\basic.exe
```
It will give you friendly errors about what dll files you are missing.

One of the problems I encountered was:
```
---------------------------
basic.exe - Entry Point Not Found
---------------------------
The procedure entry point inflateValidate could not be located in the dynamic link library C:\msys64\mingw64\bin\libpng16-16.dll. 
---------------------------
OK   
---------------------------
```
This was solved simply by upgrading the system:
```
pacman -Syu
```

Another problem that confused me was that I built a 32-bit executable, but it doesn't run with `cargo run`, but you can run it directly. I recommend just sticking to 64-bit everything on 64-bit Windows. At least that is what I'm going to do.
```
camer@delltop MINGW32 /c/Users/camer/rs/gtk-rs-examples
$ cargo run --bin basic
    Finished dev [unoptimized + debuginfo] target(s) in 0.1 secs
     Running `target\debug\basic.exe`
error: process didn't exit successfully: `target\debug\basic.exe` (exit code: 3221225785)

camer@delltop MINGW32 /c/Users/camer/rs/gtk-rs-examples
$ target/debug/basic.exe
```

**Comment by saolof** [comment]
I'm getting the same "The procedure entry point inflateValidate could not be located in the dynamic link library C:\msys64\mingw64\bin\libpng16-16.dll." error message as you, but running pacman -Syu didn't fix it.

**Comment by mati865** [comment]
@saolof  could you verify you have mingw-w64-x86_64-installed?

_Sent from my Xperia Z1 using [FastHub](https://play.google.com/store/apps/details?id=com.fastaccess.github)_

**Comment by saolof** [comment]
I have everything installed, but my install is in C:\tools\msys64 rather than C:\msys64 . That may have something to do with the problem, but I have no idea why it would try to look for libraries at the second location instead of the first.

**Comment by mati865** [comment]
MSYS2 location doesn't matter, I have it on D:\

**Comment by mathphreak** [comment]
I spent too long working on this and now it somehow works for me. I have no idea how much of this will be helpful for anybody else, but here's what worked for me.

Evidence that things were set up correctly:
- `rustup show` says my active toolchain is `stable-x86_64-pc-windows-gnu`.
- I have `mingw-w64-x86_64-gtk3` installed in MSYS2, and MSYS2 lives in `C:\msys64`.
- from a MinGW64 shell, `pkg-config --modversion gtk+-3.0` prints `3.22.26` and `pkg-config --libs gtk+-3.0` prints `-LC:/msys64/mingw64/lib` (and then a bunch of other stuff)
- `GTK_LIB_DIR` is set to `C:\msys64\mingw64\lib`
- `PATH` contains `C:\msys64\mingw64\bin`

I could compile the examples from `cmd`, but when I tried to run them I got this error:

![2018-01-05 20_50_40-basic exe - entry point not found](https://user-images.githubusercontent.com/360194/34635962-ac2fa2d0-f25c-11e7-8e9d-34b0c8be054e.png)

I could compile and run them from MinGW64 fine, though.

I still don't quite understand `ldd`, but I tried `ldd target/debug/basic.exe` from `cmd` and it referenced a bunch of .dlls from Emacs. I don't use Emacs, so I removed that folder from `%PATH%` and then I could launch the examples from `cmd`.

So if anybody is having issues and gets stuck at "entry point not found," the issue might be that your PATH has bad things in it, and `C:\msys64\usr\bin\ldd.exe target\debug\basic.exe` from a `cmd` that is broken might help you find what's wrong.

**Comment by viprip** [comment]
Good Afternoon,

When I started to using Rust for the first time, I ran `rustup-init` in order to install the `x86_64-pc-windows-msvc` toolchain. Like I would learn more about this wonderful programming language I started to read some books like *Learning Rust* and *Rust Programming by Example*. This advantage of the latter is that use both the **MSVC** and **GNU** toolchain. Starting to chapter 5, I ran into your issue and for resolving it, I searched and searched again until I figure out. The first step is to install the `x86_64-pc-windows-gnu` toolchain:

```powershell
PS C:\Users\Name\Downloads> rustup install stable-gnu
PS C:\Users\Name\Downloads> rustup default x86_64-pc-windows-gnu
```

I assume that you already have installed [MSYS 2](https://www.msys2.org/). The next step is to create a configuration file for `Cargo`:

```powershell
PS C:\Users\Name\Downloads> vim $env:USERPROFILE/.cargo/config
```

and add the following lines:

```conf
[target.x86_64-pc-windows-gnu]
linker = "D:\msys64\mingw64\bin\gcc.exe"
ar = "D:\msys64\mingw64\bin\ar.exe"
```

Now open the `MSYS2 MinGW 64-bit` console and edit the `.bashrc` in order to be abble to use `Rust` binaries:

```bash
PATH=$PATH:/d/Rust/bin
```

and reload it:

```bash
$ source ~/.bashrc
```

Go to at the root of your project and execute the following commands:

```bash
$ cargo clean
$ cargo build
```

If everything is working fine, you should be able to run `Cargo` commands within any terminals like `PowerShell` or inside the `PowerShell` terminal integrated into Visual Studio Code.

**Note:**

> I do not set the `GTK_LIB_DIR` environment variable, but I added `C:\msys64\mingw64\bin` to the `PATH` environment variable.

**Comment by mati865** [comment]
@mathphreak to run binary you just compiled you have to provide linked libraries too.

**Comment by bearmate** [comment]
I had a lot of trouble with missing entry points for `deflateSetHeader` in libgio-2.0-0.dll and `inflateReset2` in libpng16-16.dll too, but I think I've found the problem for me and possibly also for others in similar situations.

The entry points are most certainly missing, because the wrong library versions are used, because the path isn't correctly configured. I used [dependencies](https://github.com/lucasg/Dependencies) to inspect the libgio-2.0-0 and libpng16-16 libraries and found the missing symbols as a part of zlib1.dll. A quick search with [Everything](https://www.voidtools.com/) revealed at least 10 different zlib1.dll on my pc (nearly every application using gtk ships with one). After another inspection of the binary using gtk-rs with the depenecies tool I realised, that the used zlib in the dependecy graph was not the correct one from my msys64 directory even though my path was configured according to the instructions.
But I've found the path of the wrong zlib in my system path variable, which will always be evaluated before the user path (in my case it was `C:\Program Files\Intel\WiFi\bin\`).

Editing the system path (including the msys path at front or removing the wrong path) has fixed the problem.
